File: lib/pieces/horse.rb
require_relative "./piece"

class Horse < Piece

  def get_moves position
    directions.each do |direction|
      blocking_square = starting_square.send direction
      next if !blocking_square ||
              position.find(blocking_square.coordinates) ||
              !blocking_square.send(direction)
      pivot_square = blocking_square.send direction
      secondary_directions(direction).each do |secondary_direction|
        target_square = pivot_square.send secondary_direction
        next unless target_square
        blocking_piece = position.find target_square.coordinates
        next if blocking_piece && blocking_piece.color == @color
        add_move target_square.coordinates
      end
    end
  end

private 
  
  def secondary_directions direction
    if [:left, :right].include? direction
      [:up, :down]
    else
      [:left, :right]
    end
  end 
end

File: src/horse.js
var Piece = require('./piece.js');
module.exports = Horse;

function Horse(color) {
  var self = new Piece(color);

  self.type = 'Horse'

  self.getMoves = function(position) {
    var moves = [];
    var directions = {
      left: ['down', 'up'],
      right: ['down', 'up'],
      up: ['left', 'right'],
      down: ['left', 'right']
    };
    for (var pri_dir_index = 0; pri_dir_index < 4; pri_dir_index += 1) {
      var primary_direction = Object.keys(directions)[pri_dir_index];
      var blocking_square = this.square[primary_direction];
      if (!blocking_square || position[blocking_square.coordinates]) {
        continue;
      }
      var pivot_square = blocking_square[primary_direction];
      if (!pivot_square) {
        continue;
      }
      for (var sec_dir_index = 0; sec_dir_index < 2; sec_dir_index += 1) {
        var secondary_direction = directions[primary_direction][sec_dir_index];
        var target_square = pivot_square[secondary_direction];
        if (target_square && (!position[target_square.coordinates] || 
          position[target_square.coordinates].color !== this.color)) {
          moves.push(target_square.coordinates);
        }
      }
    }
    return moves;
  };

  return self;
}
